Rust: Improve CFG

paldepind opened the pull request:
With this PR CFG creation is (attempted) to cover every relevant node in the AST. Some noteworthy additions are `match` expressions, closures, `continue` and `break` in loops (including with labels), `if let` expressions and short-circuiting binary operators.

I'm not aware of any syntax constructions that are not supported, but there probably are some bugs in there :)

paldepind commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -49,25 +49,103 @@ module CfgInput implements InputSig<Location> {
+  }
+}
+
+class LogicalOrBinaryOpExprTree extends PreOrderTree instanceof BinaryExpr {
```
The short-circuiting operators are pre-ordered s.t. nested operators can be handled in a way where impossible paths are eliminated.

geoffw0 reviewed (dismissed):
Looks very promising!

> there probably are some bugs in there :)

Hopefully when we have consistency checks and actual queries depending on this we'll discover any issues quite rapidly.

paldepind commented on rust/ql/test/library-tests/controlflow/Cfg.expected:
```diff
@@ -1,92 +1,676 @@
-| test.rs:14:9:14:13 | BinaryExpr | semmle.order | 29 |
-| test.rs:14:13:14:13 | LiteralExpr | semmle.order | 30 |
+| test.rs:1:1:4:1 | enter test_call | semmle.order | 1 |
+| test.rs:1:1:4:1 | exit test_call | semmle.order | 2 |
```
Good question. I'm not 100% sure but I think that for every edge there is one row with `semmle.label` and one with `semmle.order`.

* The `semmle.label` row contains in the last column the label for the edge if it is a special kind of edge.
* The `semmle.order` row contains in the last column a number as part of an enumeration of all the edges out of a given node. I don't think the order has any meaning.

It feels like the two rows could be combined into one by adding an additional column, but there's probably a reason for it being this way 🙂

hvitved replied:
The `semmle.order` label is only relevant when using `@kind graph` as output (which is a bad idea). So I suggest changing the `getOrderDisambiguation` override to `none()`, which will remove all the unneeded `semmle.order` rows.

paldepind replied:
Doing that seems to also remove all the nodes in the file? Is that ok?

hvitved replied:
Leave it be then; I'll do it as part of https://github.com/github/codeql/pull/17520 anyway.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -117,13 +241,222 @@ class IfExprTree extends PostOrderTree instanceof IfExpr {
+    )
+    or
+    // Edge from the end of each arm to the match expression.
+    last(super.getBranch(_), pred, c) and succ = this and completionIsSimple(c)
```
`completionIsNormal` instead of `completionIsSimple`?

paldepind replied:
This change by itself causes edges to go from the arm patterns and guards to the `match` expression. I propose that we stick with `completionIsSimple` for now and then extend this later such that Boolean completions can propagate from the branch (but _not_ the guard).

paldepind replied:
This also means that we can't allow Boolean completions to be valid for the last node in a `match` arm.

hvitved replied:
That's fine, for now.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -117,13 +241,222 @@ class IfExprTree extends PostOrderTree instanceof IfExpr {
+  }
+
+  override predicate succ(AstNode pred, AstNode succ, Completion c) {
+    last(super.getExpr(), pred, c) and succ = this
```
`completionIsNormal(c)`.

hvitved reviewed (changes requested):
Really great work, Simon!

paldepind commented on rust/ql/lib/codeql/rust/controlflow/internal/SuccessorType.qll:
```diff
@@ -28,13 +40,49 @@ abstract private class ConditionalSuccessor extends SuccessorTypeImpl {
 }
 
-/** A Boolean control flow successor. */
+/** A boolean control flow successor for a boolean conditon. */
```
Why do we want to spell boolean with uppercase?

hvitved replied:
https://english.stackexchange.com/questions/4481/should-the-word-boolean-be-capitalized

paldepind replied:
TIL. Thanks :)

paldepind commented on rust/ql/lib/codeql/rust/controlflow/internal/Scope.qll:
```diff
@@ -1,10 +1,27 @@
+  override predicate scopeLast(AstNode node, Completion c) { last(this.getBody(), node, c) }
+}
+
+final class LambdaScope extends CfgScope, ClosureExpr {
```
Yes, good idea. They are indeed called closures and not lambdas in Rust 👍

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -49,25 +49,103 @@ module CfgInput implements InputSig<Location> {
 
-class FunctionTree extends StandardPostOrderTree instanceof Function {
-  override ControlFlowTree getChildNode(int i) { i = 0 and result = super.getBody() }
+/** Holds for a trivial pattern that is always guaranteed to match. */
```
Btw, here is a style guide: https://github.com/github/codeql/blob/main/docs/qldoc-style-guide.md

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -117,13 +241,222 @@ class IfExprTree extends PostOrderTree instanceof IfExpr {
+  LetStmtTree() { not trivialPat(super.getPat()) }
+
+  final override predicate propagatesAbnormal(AstNode child) {
+    child = super.getInitializer() or child = super.getElse()
```
Yes, it should do the same thing.

hvitved commented on rust/ql/lib/codeql/rust/controlflow/internal/ControlFlowGraphImpl.qll:
```diff
@@ -117,13 +241,222 @@ class IfExprTree extends PostOrderTree instanceof IfExpr {
+    // Edge from pattern to guard/arm if match succeeds.
+    pred = super.getPat() and
+    c.(MatchCompletion).succeeded() and
+    (if super.hasGuard() then first(super.getGuard(), succ) else first(super.getExpr(), succ))
```
It is fine to use `if`, but only if the condition is not implied in the `then` branch (and anti-implied in the `else` branch).

paldepind replied:
I see. Makes sense :)

paldepind commented on rust/ql/lib/codeql/rust/controlflow/internal/Completion.qll:
```diff
@@ -6,7 +7,10 @@ private import SuccessorType
   TBooleanCompletion(Boolean b) or
-  TReturnCompletion()
+  TMatchCompletion(Boolean isMatch) or
+  TLoopCompletion(TLoopJumpType kind, TLabelType label) or
```
I've tried to do this. Please double check that it looks right :)

hvitved replied:
It does :-)

paldepind commented:
The PR is ready for re-review.

I've addressed most of the comments. I didn't fix the `match`/`no-match` edges going out of `let` statements. I don't understand yet why they're not picked up by the `trivialPat` predicate, but we can investigate that later I think :)

hvitved commented:
> The PR is ready for re-review.

I don't understand why the CFGs for `test_nested_if` and `test_nested_if_match` are gone in the merge commit?

paldepind commented:
You're right, I didn't notice that. It could be due to the changes in the extractor or maybe some of the last changes I made broke it. Let's investigate after merging :)
